### PR TITLE
EZP-30978: fix issue when location limitation is used on content/edit policy

### PR DESCRIPTION
**Pull request by erdnaxelaweb**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-30978](https://jira.ez.no/browse/EZP-30978)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** | `7.x`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

https://jira.ez.no/browse/EZP-30978

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [ ] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [ ] Ask for Code Review.

### CODE REVIEW

File: eZ/Publish/Core/Limitation/LocationLimitationType.php [hunk 1]
```diff
@@ -125,25 +128,33 @@ public function evaluate(APILimitationValue $value, APIUserReference $currentUse
         } elseif ($object instanceof VersionInfo) {
             $object = $object->getContentInfo();
         } elseif (!$object instanceof ContentInfo) {
-            throw new InvalidArgumentException(
-                '$object',
-                'Must be of type: ContentCreateStruct, Content, VersionInfo or ContentInfo'
-            );
+            // As this is Role limitation we need to signal abstain on unsupported $object
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Limitation/LocationLimitationType.php]: Only SectionLimitation and SubtreeLimitation are available when you assign a Role to the User or User Group. I am not saying that we should not return `ACCESS_ABSTAIN` but not for this reason. Also, this change is not needed to solve this issue. I am in favor of not making this change now.

File: eZ/Publish/Core/Limitation/LocationLimitationType.php [hunk 2]
```diff
@@ -125,25 +128,33 @@ public function evaluate(APILimitationValue $value, APIUserReference $currentUse
...
-                // @todo Need support for draft locations to to work correctly
+                // @todo Need support for draft locations to work correctly
                 $targets = $this->persistence->locationHandler()->loadParentLocationsForDraftContent($object->id);
             }
         }
 
         foreach ($targets as $target) {
             if (!$target instanceof Location && !$target instanceof SPILocation) {
-                throw new InvalidArgumentException('$targets', 'Must contain objects of type: Location');
+                // As this is Role limitation we need to signal abstain on unsupported $targets
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Limitation/LocationLimitationType.php]: The same here. LocationLimitation is not available when you assign a Role to the User or User Group. We get rid off `Versions` from `$targets`. I think that this change is not needed or at least explanation should be different.

**Review by mikadamczyk — CHANGES_REQUESTED** [review]

**Comment by florianalexandre** [comment]
I have gone a bit overboard after taking the subtree limitation as an exemple.
Reverted some modification to only keep what is need to fix the original issue.

**Review by mikadamczyk — APPROVED** [review]

File: eZ/Publish/Core/Limitation/LocationLimitationType.php [hunk 3]
```diff
@@ -131,12 +134,16 @@ public function evaluate(APILimitationValue $value, APIUserReference $currentUse
             );
         }
 
+        $targets = array_filter($targets, function ($target) {
+            return !$target instanceof Version;
```
Comment (alongosz) [inline eZ/Publish/Core/Limitation/LocationLimitationType.php]: The actual bug exists in the implementation of `\eZ\Publish\Core\Repository\Permission\PermissionResolver::lookupLimitations`. 
An instance of `Target` should be passed to `TargetAwareType` Limitation type only, which [`canUser` takes into the account](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.4/eZ/Publish/Core/Repository/Permission/PermissionResolver.php#L224). The `lookupLimitation` method needs to behave the same way. // cc @mikadamczyk 

While this fix works for `LocationLimitation`, there might be similar issues 

Ref:
- `\eZ\Publish\SPI\Limitation\TargetAwareType`,
- `\eZ\Publish\SPI\Limitation\Target`.

**Review by alongosz — CHANGES_REQUESTED** [review]
Please add an integration test for the use case so we can all understand what is the setup which causes this issue. It can be created either in `\eZ\Publish\API\Repository\Tests\PermissionResolverTest` or in `\eZ\Publish\API\Repository\Tests\Values\User\Limitation\LocationLimitationTest`.

Moreover the solution needs to be changed so it applies to all Limitations (in a generic way), not just Location:

**Comment by adamwojs** [comment]
Closing PR as obsolete. eZ Platform 2.5 has reached EOM, so please reopen PR in ezsystems/ezplatform-kernel (bug fixes) or ibexa/core (features/improvements) if issue is still valid for v3.3 / v4.x and you are willing to work on it.

**Pull request closed without merging**